File: api/ti99_.cxx
/**
 *  Java Grinder
 *  Author: Michael Kohn
 *   Email: mike@mikekohn.net
 *     Web: http://www.mikekohn.net/
 * License: GPLv3
 *
 * Copyright 2014-2018 by Michael Kohn
 *
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdint.h>

#include "JavaClass.h"
#include "ti99_.h"

#define CHECK_FUNC(funct,sig) \
  if (strcmp(#funct#sig, method_name) == 0) \
  { \
    return generator->ti99_##funct##sig(); \
  }

#define CHECK_FUNC_CONST(funct,sig) \
  if (strcmp(#funct#sig, method_name) == 0) \
  { \
    return generator->ti99_##funct##sig(const_val); \
  }

#define CHECK_FUNC_CONST_2(funct,sig) \
  if (strcmp(#funct#sig, function) == 0) \
  { \
    return generator->ti99_##funct##sig(const_val1, const_val2); \
  }

int ti99(JavaClass *java_class, Generator *generator, const char *method_name)
{
  CHECK_FUNC(print,_X)
  CHECK_FUNC(printChar,_C)
  CHECK_FUNC(setCursor,_II)
  CHECK_FUNC(setGraphicsMode,_I)
  CHECK_FUNC(clearScreen,)
  CHECK_FUNC(plot,_III)
  CHECK_FUNC(setColors,)
  CHECK_FUNC(setSoundFreq,_II)
  CHECK_FUNC(setSoundVolume,_II)
  CHECK_FUNC(setSpriteVisible,_IZ)
  CHECK_FUNC(setSpriteImage,_IaB)
  CHECK_FUNC(setSpritePos,_III)
  CHECK_FUNC(setSpriteColor,_II)
  CHECK_FUNC(setSpriteSize,_I)

  return -1;
}

int ti99(JavaClass *java_class, Generator *generator, const char *method_name, int const_val)
{
  CHECK_FUNC_CONST(printChar,_C)
  CHECK_FUNC_CONST(setGraphicsMode,_I)
  //CHECK_FUNC_CONST(setSoundFreq,_II)
  //CHECK_FUNC_CONST(setSoundVolume,_II)

  return -1;
}

int ti99(JavaClass *java_class, Generator *generator, const char *function, int const_val1, int const_val2)
{
  CHECK_FUNC_CONST_2(setCursor,_II)

  return -1;
}

File: java/net/mikekohn/java_grinder/TI99.java
/**
 *  Java Grinder
 *  Author: Michael Kohn
 *   Email: mike@mikekohn.net
 *     Web: http://www.naken.cc/
 * License: GPL
 *
 * Copyright 2014-2015 by Michael Kohn
 *
 */

package net.mikekohn.java_grinder;

abstract public class TI99
{
  public static final int SPRITE_SIZE_8X8_SMALL = 0;
  public static final int SPRITE_SIZE_8X8_BIG = 1;
  public static final int SPRITE_SIZE_16X16_SMALL = 2;
  public static final int SPRITE_SIZE_16X16_BIG = 3;

  /** Standard Mode (Text) (32x24) */
  public static final int MODE_0 = 0;
  /** Text Mode (40x24) */
  public static final int MODE_1 = 1;
  /** Multicolor mode (64x48) */
  public static final int MODE_2 = 2;
  /** Bitmap mode (?) */
  public static final int MODE_3 = 3;

  protected TI99() { }

  //public static void clearScreen() { }
  public static void print(String text) { }
  public static void printChar(char c) { }
  public static void setCursor(int x, int y) { }
  public static void setGraphicsMode(int mode) { }
  public static void clearScreen() { }
  public static void plot(int x, int y, int color) { }

  /** For mode 0, remove characters and set colors for drawing */
  public static void setColors() { }

  /** Set sound frequency for a sound channel (voice).  The voice
      param can be between 0 to 2 for normal notes and 3 for a noise
      voice.  Anything out of that range will cause unknown results.
      The freq param is based on the table from this page:
      http://www.nouspikel.com/ti99/titechpages.htm */
  public static void setSoundFreq(int voice, int freq) { }

  /** Set sound volume for a sound channel (voice).  The voice
      param can be between 0 to 2 for normal notes and 3 for a noise
      voice.  Anything out of that range will cause unknown results.
      The volume can be between 0 and 15.  0 = loudest, 15 = sound is off. */
  public static void setSoundVolume(int voice, int volume) { }

  /** Change visibility of a sprite.  Sprite index can be between 0 and 32.
      there are restrictions on the number of sprites that can be displayed
      on a line. */
  public static void setSpriteVisible(int index, boolean visible) { }

  /** Sets what a sprite looks like.  Takes an array of 8 or 32 bytes. */
  public static void setSpriteImage(int index, byte[] image) { }

  /** Sets the position on the screen where a sprite is seen. */
  public static void setSpritePos(int index, int x, int y) { }

  /** Sets the color (between 0 and 15) of the sprite. */
  public static void setSpriteColor(int index, int color) { }

  /** Sets if the sprite is 8x8 or 16x16 pixels. */
  public static void setSpriteSize(int size) { }
}

